### PR TITLE
Hide all messaging that includes URLs

### PR DESCRIPTION
**Pull request by bryceitoc9**
## Description
* Doubles-down on using the `HttpResourceFetcher` class to make all requests to external URLs that aren't a part of an SDK call.
  * Puts all code using the deprecated `request` library in a single place. This will be removed at some point and will make it easier to track down.
  * Hides all URLs from errors thrown by the `request` library.
  * Forces `HttpResourceFetcher` to opt-in to showing external URLs when logging. This is typically useful, but will force devs to think twice about whether or not to show a potentially private URL.

## Motivation and Context

Suggested by our security reviewers, especially tied to Lambda import functionality. Specifically, Lambda import uses the AWS Lambda's [`getFunction` call](https://docs.aws.amazon.com/lambda/latest/dg/API_GetFunction.html), which produces a presigned link to a user's Lambda function's source which is valid for 10 minutes. This URL provides no value to our logs but does produce a vector through which malicious users can steal the function's code.

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the changelog using the script `npm run newChange`

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/resourcefetcher/httpResourceFetcher.ts [hunk 1]
```diff
@@ -36,7 +36,8 @@ export class HttpResourceFetcher implements ResourceFetcher {
         return new Promise<string>((resolve, reject) => {
             request(this.url, (error: any, response: IncomingMessage, body: any) => {
                 if (error) {
-                    reject(error)
+                    // Throw sanitized error so we don't end up revealing any potentially private URLs.
+                    reject(new Error('Error loading from URL'))
```
Comment (awschristou) [inline src/shared/resourcefetcher/httpResourceFetcher.ts:40]: Will fully obscuring the error make it harder to diagnose issues? The error may contain other attributes (like an error code) which calling code may want to react to. Is it feasible to obscuring just the url from the error message, but otherwise leave the error object as-is?

**Review by awschristou — COMMENTED** [review]
This looks like an important change based on the PR description, but there are no tests. How can future regressions be prevented (whether someone is writing all-new code, or modifying code in the areas this PR changes)?

File: src/shared/resourcefetcher/httpResourceFetcher.ts (hunk 1 above)
Comment (bryceitoc9) [reply src/shared/resourcefetcher/httpResourceFetcher.ts:40]: I haven't done a deep dive on where the URL can surface in the error message. Feels kinda hacky to remove/alter fields from the request object but I can think about it. I'm not 100% sure what information we'd get from a failed request though.

**Comment by bryceitoc9** [comment]
> This looks like an important change based on the PR description, but there are no tests. How can future regressions be prevented (whether someone is writing all-new code, or modifying code in the areas this PR changes)?

I think the only way to ensure this would be to turn this into a separate helper function that wraps `request` or some other library. It's worth considering but this is the lightest-weight thing to do right now to address the immediate issue. Would you rather me move in that direction?

File: src/lambda/commands/importLambda.ts [hunk 2]
```diff
@@ -155,7 +155,8 @@ async function downloadAndUnzipLambda(
                     resolve()
                 })
                 .on('error', err => {
-                    throw err
+                    // Throw sanitized error so we don't reveal a potentially active URL to private resources.
+                    throw new Error('Error downloading Lambda code from AWS.')
```
Comment (justinmk3) [inline src/lambda/commands/importLambda.ts:159]: mentioning the name or arn should be ok though.
```suggestion
                    throw new Error(`Error downloading Lambda code from AWS: ${functionArn}`)
```

**Review by justinmk3 — APPROVED** [review]

**Comment by bryceitoc9** [comment]
> 
> 
> > This looks like an important change based on the PR description, but there are no tests. How can future regressions be prevented (whether someone is writing all-new code, or modifying code in the areas this PR changes)?
> 
> I think the only way to ensure this would be to turn this into a separate helper function that wraps `request` or some other library. It's worth considering but this is the lightest-weight thing to do right now to address the immediate issue. Would you rather me move in that direction?

@awschristou Moved everything to common functions (not everything was using the `HttpResourceFetcher` when they should have been!), I'm not sure if there's a great way to test this however. Common code throughout should help with regressions.

Can add tests if you insist (FWIW HttpResourceFetcher has none at this time)

File: src/shared/utilities/requestUtils.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import * as request from 'request'
+
+/**
+ * Makes a `get` request to a URL without params and returns the response asynchronously.
+ * Can optionally pipe the call's data to a location.
+ * Swallows error to hide the targeted URL. Recommended by security.
+ * @param url URL to send request to
+ * @param pipeLocation Filepath to pipe output to.
+ */
+export async function getResponseFromGetRequest(url: string, pipeLocation?: string): Promise<request.Response> {
```
Comment (justinmk3) [inline src/shared/utilities/requestUtils.ts:17]: could live in `HttpResourceFetcher`. Either `HttpResourceFetcher` or this requestUtils is redundant. Better to keep related code together than spread it out in many tiny files.

this also looks very similar to the purpose of `HttpResourceFetcher`.

Comment (bryceitoc9) [reply src/shared/utilities/requestUtils.ts:17]: This one is specifically to accommodate the pipe. I'll play with it a little more to see if I can wire that up there.

File: src/shared/resourcefetcher/httpResourceFetcher.ts [hunk 4]
```diff
@@ -3,46 +3,41 @@
...
-                if (error) {
-                    // Throw sanitized error so we don't end up revealing any potentially private URLs.
-                    reject(new Error('Error loading from URL'))
-                } else {
-                    // tslint:disable-next-line: no-unsafe-any
-                    resolve(body.toString())
-                }
-            })
-        })
+        return (await getResponseFromGetRequest(this.url)).body
```
Comment (justinmk3) [inline src/shared/resourcefetcher/httpResourceFetcher.ts:41]: HttpResourceFetcher does close to nothing now. If we are trying to avoid changing a lot of call-sites, we can at least keep the new "util" function in HttpResourceFetcher .

File: src/shared/resourcefetcher/httpResourceFetcher.ts [hunk 5]
```diff
@@ -3,46 +3,41 @@
...
+            this.logger.error(`Error loading ${this.logText()}: %O`, err as Error)
 
             return undefined
         }
     }
 
+    // TODO: Are there cases where we don't mind the URL?
+    // Safer to do it this way assuming others use this library.
+    private logText(): string {
+        return this.friendlyName ?? 'resource from external URL'
```
Comment (justinmk3) [inline src/shared/resourcefetcher/httpResourceFetcher.ts:37]: what does "external" mean here? is there anything preventing use of an "internal" URL with this module?

File: src/shared/utilities/requestUtils.ts (hunk 3 above)
Comment (justinmk3) [reply src/shared/utilities/requestUtils.ts:17]: it could still live in `HttpResourceFetcher` (as a module function or a static class function) or whatever module we want to keep HTTP stuff in.

File: src/shared/resourcefetcher/httpResourceFetcher.ts [hunk 6]
```diff
@@ -3,45 +3,41 @@
...
-import * as request from 'request'
 import { getLogger, Logger } from '../logger'
+import { getResponseFromGetRequest } from '../utilities/requestUtils'
 import { ResourceFetcher } from './resourcefetcher'
 
 export class HttpResourceFetcher implements ResourceFetcher {
     private readonly logger: Logger = getLogger()
 
-    public constructor(private readonly url: string) {}
+    public constructor(private readonly url: string, private readonly friendlyName?: string) {}
```
Comment (awschristou) [inline src/shared/resourcefetcher/httpResourceFetcher.ts:13]: This change makes all URLs opaque, which is not how some of the existing usage is intended. One example is when loading endpoints data: 

```
2020-10-27 08:36:19 [VERBOSE]: Loading resource from https://idetoolkits.amazonwebservices.com/endpoints.json
2020-10-27 08:36:20 [VERBOSE]: Finished loading resource from https://idetoolkits.amazonwebservices.com/endpoints.json
```

It should be possible for callers to use `HttpResourceFetcher` in a way that displays or masks the url, at the caller's discretion.

**Review by awschristou — CHANGES_REQUESTED** [review]

File: src/shared/resourcefetcher/httpResourceFetcher.ts (hunk 6 above)
Comment (bryceitoc9) [reply src/shared/resourcefetcher/httpResourceFetcher.ts:13]: I'm not sure how this makes it opaque? The URL is a constant and the key is a constant as well. Is there that much value in showing any of the URLs that I have masked?

File: src/shared/resourcefetcher/httpResourceFetcher.ts (hunk 5 above)
Comment (bryceitoc9) [reply src/shared/resourcefetcher/httpResourceFetcher.ts:37]: Just a filler word. If you have a better phrase feel free to suggest.

File: src/shared/resourcefetcher/httpResourceFetcher.ts (hunk 6 above)
Comment (awschristou) [reply src/shared/resourcefetcher/httpResourceFetcher.ts:13]: If I understand the change, this class will now never emit a url to logs. This means all existing usages where urls were intended to be written to logs are affected, including the example above. There are places where providing a URL makes it easier to understand/troubleshoot/diagnose what is happening, particularly if/when issues arise in the toolkit. With the endpoints example, this provides a way to understand where the Toolkit is connecting to, and try to reproduce any download failures outside the toolkit.

Comment (bryceitoc9) [reply src/shared/resourcefetcher/httpResourceFetcher.ts:13]: So as of now, this change masks the following URLs:
* Endpoint XML
* Lambda sample request manifest
* Lambda getFunction URL (must mask)
* VSDBG install script
* SFN webview assets (JS/CSS)

I'm not convinced there's a lot of value in surfacing any of those URLs, again, since they're all constants and we can find the location easily. Is the concern that someone would have the URL blocked by a firewall rule that they'd have to override?

At the very least, surfacing the URL will have to be an opt-in. I'd rather make showing the URL a conscious choice to help short-circuit some of these issues in the future. Maybe have a mandatory `isUserData` param in the resource fetcher?

Comment (awschristou) [reply src/shared/resourcefetcher/httpResourceFetcher.ts:13]: * Endpoint XML, Lambda sample request manifest, VSDBG install script
These URLs are verbose level information that articulates what the toolkit is doing, and help to troubleshoot in relevant circumstances. They give users a chance to self-diagnose where applicable ("oh hey, I have that domain blocked"), and make it easier for us to assist when logs are shared in issues. They are constant values today, but could change for reasons we don't know today. Making them opaque would require knowledge of which version is running (which we might not have), in order to look up the corresponding value. Logging the URL avoids those additional lookup steps.

* Lambda getFunction URL / SFN webview assets (JS/CSS)
These are the URLs you're interested in masking with this PR

Some options are:
* Maintain the pre-PR behavior of `HttpResourceFetcher` as the default behavior (ie: not backwards incompatible). Add an optional configuration object to the constructor, which can mask the URL, and optionally provide a mask text.
* Leave HttpResourceFetcher as-is, and add a class that derives from it, which does URL masking.

Comment (bryceitoc9) [reply src/shared/resourcefetcher/httpResourceFetcher.ts:13]: > *    Maintain the pre-PR behavior of HttpResourceFetcher as the default behavior (ie: not backwards incompatible). Add an optional configuration object to the constructor, which can mask the URL, and optionally provide a mask text.
> *    Leave HttpResourceFetcher as-is, and add a class that derives from it, which does URL masking.

I _really_ don't like the idea of having it show the URL by default; it should be opt-in so the dev has to think about whether or not it should be shown. The change itself would be minimal...it would be similar to the change I put here with the friendlyName param.

>Lambda getFunction URL / SFN webview assets (JS/CSS)
>These are the URLs you're interested in masking with this PR

Why would SFN fall into this category?

Comment (awschristou) [reply src/shared/resourcefetcher/httpResourceFetcher.ts:13]: > Why would SFN fall into this category?

Whoops, that's a mistake. I'll edit the comment

File: src/shared/resourcefetcher/httpResourceFetcher.ts (hunk 5 above)
Comment (justinmk3) [reply src/shared/resourcefetcher/httpResourceFetcher.ts]: i would remove "external"

Comment (bryceitoc9) [reply src/shared/resourcefetcher/httpResourceFetcher.ts]: Already done

File: src/shared/resourcefetcher/httpResourceFetcher.ts (hunk 4 above)
Comment (bryceitoc9) [reply src/shared/resourcefetcher/httpResourceFetcher.ts]: Moved

File: src/shared/utilities/requestUtils.ts (hunk 3 above)
Comment (bryceitoc9) [reply src/shared/utilities/requestUtils.ts]: Moved

File: src/lambda/commands/importLambda.ts (hunk 2 above)
Comment (bryceitoc9) [reply src/lambda/commands/importLambda.ts]: Irrelevant; handled in the catch statement further down in the file.

File: src/shared/resourcefetcher/httpResourceFetcher.ts (hunk 1 above)
Comment (bryceitoc9) [reply src/shared/resourcefetcher/httpResourceFetcher.ts]: In a stunning turn of events, I made it possible to show the URL and also did a hacky thing. Score one for @awschristou .

File: src/shared/resourcefetcher/httpResourceFetcher.ts [hunk 7]
```diff
@@ -3,45 +3,59 @@
...
 
-    private async loadFromUrl(): Promise<string | undefined> {
-        return new Promise<string>((resolve, reject) => {
-            request(this.url, (error: any, response: IncomingMessage, body: any) => {
-                if (error) {
-                    reject(error)
-                } else {
-                    // tslint:disable-next-line: no-unsafe-any
-                    resolve(body.toString())
+    // TODO: Are there cases where we don't mind the URL?
```
Comment (awschristou) [inline src/shared/resourcefetcher/httpResourceFetcher.ts:39]: this todo can be removed now

File: src/shared/resourcefetcher/httpResourceFetcher.ts [hunk 8]
```diff
@@ -3,45 +3,59 @@
...
 import { getLogger, Logger } from '../logger'
 import { ResourceFetcher } from './resourcefetcher'
 
 export class HttpResourceFetcher implements ResourceFetcher {
     private readonly logger: Logger = getLogger()
 
-    public constructor(private readonly url: string) {}
+    public constructor(
+        private readonly url: string,
+        private readonly params: { showUrl: boolean; friendlyName?: string; pipeLocation?: string }
```
Comment (awschristou) [inline src/shared/resourcefetcher/httpResourceFetcher.ts:17]: Consider adding comments to explain pipeLocation

File: src/shared/resourcefetcher/httpResourceFetcher.ts [hunk 9]
```diff
@@ -3,45 +3,59 @@
...
+        return this.params.showUrl ? this.url : this.params.friendlyName ?? 'resource from URL'
+    }
+
+    private async getResponseFromGetRequest(): Promise<request.Response> {
+        return new Promise<request.Response>((resolve, reject) => {
+            const call = request(this.url, (err, response, body) => {
+                if (err) {
+                    // swallow error to keep URL private
+                    // some AWS APIs use presigned links (e.g. Lambda.getFunction); showing these represent a securty concern.
+                    reject('Error making request to URL.')
```
Comment (awschristou) [inline src/shared/resourcefetcher/httpResourceFetcher.ts:51]: err is eaten, consider outputting err.code (if available) or relevant but safe status fields in the reject call

**Review by awschristou — APPROVED** [review]

File: src/shared/resourcefetcher/httpResourceFetcher.ts (hunk 9 above)
Comment (bryceitoc9) [reply src/shared/resourcefetcher/httpResourceFetcher.ts:51]: Code might work. Message and stack both contain the URL.

File: src/shared/clients/defaultLambdaClient.ts [hunk 10]
```diff
@@ -62,7 +62,11 @@ export class DefaultLambdaClient implements LambdaClient {
 
         try {
             const response = await client.getFunction({ FunctionName: name }).promise()
-            getLogger().debug('GetFunction returned response: %O', response)
+            // prune `Code` from logs so we don't reveal a signed link to customer resources.
+            getLogger().debug('GetFunction returned response (code section pruned): %O', {
+                ...response,
+                Code: 'Pruned',
+            })
```
Comment (justinmk3) [inline src/shared/clients/defaultLambdaClient.ts:69]: seems extreme that we can't show debug information in `debug()` logs ...

Comment (bryceitoc9) [reply src/shared/clients/defaultLambdaClient.ts:69]: It's extreme but it also doesn't provide much information (aside from maybe the `RepositoryType`?: https://docs.aws.amazon.com/lambda/latest/dg/API_FunctionCodeLocation.html ); at worst, it represents a security concern and at least provides little value because the link will be inactive after 10 minutes.

We can check internally to see whether or not this is necessary.

**Review by justinmk3 — APPROVED** [review]
